### PR TITLE
Prepare port runtime data storage for multiple ports

### PR DESCRIPTION
**Pull request by pyhys**
Implement pf_port.c file for the data retrieval

### CODE REVIEW

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]
Looking good!

File: src/device/pf_port.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ */
+int pf_port_get_next_port (pf_port_iterator_t * p_iterator);
+
+/**
+ * Get a reference to port runtime data.
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param loc_port_num     In:    Local port number.
+ *                                Valid range: 1 .. PNET_MAX_PORT
+ * @param pp_port_data     Out:   Port runtime data, or NULL if not found.
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.h:63]: Good suggestion,
I will change to this behavior.

[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.h]: Done

File: src/device/pf_port.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * @param p_iterator       InOut: Port iterator.
+ * @return Local port number for next port on local interface.
+ *         If no more ports are available, 0 is returned.
+ */
+int pf_port_get_next_port (pf_port_iterator_t * p_iterator);
+
+/**
+ * Get a reference to port runtime data.
+ *
```
Comment (olbjo) [inline src/device/pf_port.h:59]:
```suggestion
 * If the local port number is out of range this operation will assert. 
 * NULL will never be returned.
```

**Review by olbjo — APPROVED** [review]
Looks great!
I think a comment on pf_port_get_port_data() describing that it never will return NULL is useful

File: src/device/pf_port.h (hunk 2 above)
Comment (pyhys) [reply src/device/pf_port.h:59]: Done

File: src/device/pf_pdport.c [hunk 3]
```diff
@@ -51,12 +52,17 @@ static int pf_pdport_load (pnet_t * net)
          "PDPORT(%d): Did read PDPort settings from nvm.\n",
          __LINE__);
 
-      memcpy (&net->port[0].check.peer, &file_port, sizeof (pnet_pdport_nvm_t));
+      p_port_data = pf_port_get_port_data (net, PNET_PORT_1);
+
+      memcpy (
+         &p_port_data->pdport.check.peer,
+         &file_port,
+         sizeof (pnet_pdport_nvm_t));
```
Comment (fredrikdanielmoller) [inline src/device/pf_pdport.c:60]:
```suggestion
      p_port_data->pdport.check.peer = file_port.peer;
```
Note that p_port_data->pdport.check.peer is not the same type as pnet_pdport_nvm_t.

File: src/device/pf_pdport.c [hunk 4]
```diff
@@ -77,10 +83,13 @@ static int pf_pdport_save (pnet_t * net)
    pnet_pdport_nvm_t output_port_nvm = {0};
    pnet_pdport_nvm_t temporary_buffer;
    const char * p_file_directory = NULL;
+   pf_port_t * p_port_data = NULL;
+
+   p_port_data = pf_port_get_port_data (net, PNET_PORT_1);
```
Comment (fredrikdanielmoller) [inline src/device/pf_pdport.c:88]:
```suggestion
   const pf_port_t * p_port_data = pf_port_get_port_data (net, PNET_PORT_1);
```

File: src/device/pf_pdport.c [hunk 5]
```diff
@@ -152,10 +161,13 @@ int pf_pdport_init (pnet_t * net)
 int pf_pdport_reset_all (pnet_t * net)
 {
    const char * p_file_directory = NULL;
+   pf_port_t * p_port_data = NULL;
```
Comment (fredrikdanielmoller) [inline src/device/pf_pdport.c:164]:
```suggestion
   pf_port_t * p_port_data = pf_port_get_port_data (net, PNET_PORT_1);
```

File: src/device/pf_pdport.c [hunk 6]
```diff
@@ -152,10 +161,13 @@ int pf_pdport_init (pnet_t * net)
...
 
    (void)pf_cmina_get_file_directory (net, &p_file_directory);
    pf_file_clear (p_file_directory, PNET_FILENAME_PDPORT);
-   memset (&net->port[0], 0, sizeof (net->port[0]));
+
+   p_port_data = pf_port_get_port_data (net, PNET_PORT_1);
```
Comment (fredrikdanielmoller) [inline src/device/pf_pdport.c:169]: See above

File: src/common/pf_lldp.c [hunk 7]
```diff
@@ -438,7 +438,14 @@ static void pf_lldp_receive_timeout (
    void * arg,
    uint32_t current_time)
 {
-   net->lldp_rx_timeout = 0;
+   pf_port_t * p_port_data = NULL;
```
Comment (fredrikdanielmoller) [inline src/common/pf_lldp.c:441]:
```suggestion
   pf_port_t * p_port_data = pf_port_get_port_data (net, PNET_PORT_1);
```

File: src/common/pf_lldp.c [hunk 8]
```diff
@@ -438,7 +438,14 @@ static void pf_lldp_receive_timeout (
...
+
+   /* TODO Make sure the scheduling works for multiple ports
+      Find out which port is timing out. */
+
+   p_port_data = pf_port_get_port_data (net, PNET_PORT_1);
```
Comment (fredrikdanielmoller) [inline src/common/pf_lldp.c:446]: See above

File: src/common/pf_lldp.c [hunk 9]
```diff
@@ -1061,7 +1036,14 @@ void pf_lldp_tx_restart (pnet_t * net, bool send)
 
 void pf_lldp_init (pnet_t * net)
 {
-   memset (&net->lldp_peer_info, 0, sizeof (net->lldp_peer_info));
+   uint16_t loc_port_num;
+   pf_port_t * p_port_data = NULL;
+
+   for (loc_port_num = 1; loc_port_num <= PNET_MAX_PORT; loc_port_num++)
```
Comment (fredrikdanielmoller) [inline src/common/pf_lldp.c:1042]: Use the port iterator to iterate over ports. See pf_port_init_port_iterator().

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

File: src/common/pf_lldp.c (hunk 9 above)
Comment (pyhys) [reply src/common/pf_lldp.c:1042]: Nice!

Comment (pyhys) [reply src/common/pf_lldp.c]: Done

File: src/common/pf_lldp.c (hunk 7 above)
Comment (pyhys) [reply src/common/pf_lldp.c]: Done

File: src/common/pf_lldp.c (hunk 8 above)
Comment (pyhys) [reply src/common/pf_lldp.c]: Done

File: src/device/pf_port.h [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+ * If the local port number is out of range this operation will assert.
+ * NULL will never be returned.
+ *
+ * @param net              InOut: The p-net stack instance
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.h]: Done

File: src/device/pf_pdport.c (hunk 3 above)
Comment (pyhys) [reply src/device/pf_pdport.c]: Done

File: src/device/pf_pdport.c (hunk 4 above)
Comment (pyhys) [reply src/device/pf_pdport.c]: Done

File: src/device/pf_pdport.c (hunk 5 above)
Comment (pyhys) [reply src/device/pf_pdport.c]: Done

File: src/device/pf_pdport.c (hunk 6 above)
Comment (pyhys) [reply src/device/pf_pdport.c]: Done

File: test/test_port.cpp [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+TEST_F (PortUnitTest, PortCheckItetator)
+{
+   pnet_t netstruct; /* Do not care about actual value */
+   pf_port_iterator_t port_iterator;
+   int port;
+   int ix;
+
+   /* Check precondiiton. Ports must be enabled for the stack to be working. */
+   EXPECT_TRUE (PNET_MAX_PORT > 0);
```
Comment (fredrikdanielmoller) [inline test/test_port.cpp:35]: Remove this. That PNET_MAX_PORT is at least 1 should be enforced at compile-time.

File: test/test_port.cpp [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+#include <gtest/gtest.h>
+
+class PortUnitTest : public PnetUnitTest
+{
+};
+
+TEST_F (PortUnitTest, PortCheckItetator)
+{
+   pnet_t netstruct; /* Do not care about actual value */
```
Comment (fredrikdanielmoller) [inline test/test_port.cpp:29]: netstruct should be initialised. In the future, pf_port_init_port_iterator() might be changed to access the contents of it.

File: test/test_port.cpp [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+   /* Verify that we return 0 when we are done */
+   port = pf_port_get_next_port (&port_iterator);
+   EXPECT_EQ (port, 0);
+   port = pf_port_get_next_port (&port_iterator);
+   EXPECT_EQ (port, 0);
+   port = pf_port_get_next_port (&port_iterator);
+   EXPECT_EQ (port, 0);
+   port = pf_port_get_next_port (&port_iterator);
+   EXPECT_EQ (port, 0);
```
Comment (fredrikdanielmoller) [inline test/test_port.cpp:57]: Remove these as testing the same thing four times is a bit excessive.

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

File: src/device/pf_port.c [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+void pf_port_get_port_list (pnet_t * net, pf_lldp_port_list_t * p_list)
+{
+   /* TODO: Implement support for multiple ports */
+   memset (p_list, 0, sizeof (*p_list));
+   p_list->ports[0] = BIT (8 - 1);
+}
+
+void pf_port_init_port_iterator (pnet_t * net, pf_port_iterator_t * p_iterator)
+{
+   p_iterator->next_port = (PNET_MAX_PORT > 0) ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.c:34]: Done

File: src/device/pf_port.c [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+int pf_port_get_next_port (pf_port_iterator_t * p_iterator)
+{
+   int ret = p_iterator->next_port;
+
+   if (p_iterator->next_port >= PNET_MAX_PORT || p_iterator->next_port == 0)
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.c:41]: Changed to 
`if (p_iterator->next_port == PNET_MAX_PORT || p_iterator->next_port == 0)`
in order to not restart automatically.
Done.

File: src/device/pf_port.h [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+ *
+ * If the local port number is out of range this operation will assert.
+ * NULL will never be returned.
+ *
+ * @param net              In:    The p-net stack instance
+ * @param loc_port_num     In:    Local port number.
+ *                                Valid range: 1 .. PNET_MAX_PORT
+ * @return Address to port runtime data
+ */
+pf_port_t * pf_port_get_port_data (
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.h:68]: Done

File: src/device/pf_port.h [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Get next local port.
+ *
+ * If no more ports are available, 0 is returned.
+ *
+ * @param p_iterator       InOut: Port iterator.
+ * @return Local port number for next port on local interface.
+ *         If no more ports are available, 0 is returned.
+ */
+int pf_port_get_next_port (pf_port_iterator_t * p_iterator);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.h:55]: Done

File: src/device/pf_port.h [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+ * Initialise iterator for iterating over local ports.
+ *
+ * This iterator may be used for iterating over all physical ports
+ * on the local interface. The management port is not included.
+ * See pf_port_get_next_port().
+ *
+ * @param net              In:    The p-net stack instance.
+ * @param p_iterator       Out:   Port iterator.
+ */
+void pf_port_init_port_iterator (pnet_t * net, pf_port_iterator_t * p_iterator);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.h:44]: Done

File: src/device/pf_port.h [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Get list of local ports.
+ *
+ * This is the list of physical ports on the local interface.
+ * The management port is not included.
+ *
+ * @param net              In:    The p-net stack instance.
+ * @param p_list           Out:   List of local ports.
+ */
+void pf_port_get_port_list (pnet_t * net, pf_lldp_port_list_t * p_list);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_port.h:32]: Done

File: test/test_port.cpp (hunk 11 above)
Comment (pyhys) [reply test/test_port.cpp:35]: Done

File: test/test_port.cpp (hunk 12 above)
Comment (pyhys) [reply test/test_port.cpp:29]: Using the integration test base class instead, where `net` is initialized.
Done

File: test/test_port.cpp (hunk 13 above)
Comment (pyhys) [reply test/test_port.cpp:57]: Done

**Review by fredrikdanielmoller — APPROVED** [review]

**Pull request merged**